kernel/mmanutils.c

```diff
@@ -780,59 +902,31 @@ int myst_msync(void* addr, size_t length, int flags)
     if ((flags & MS_SYNC) && (flags & MS_ASYNC))
         ERAISE(-EINVAL);
 
-    // Note: Asynchronous sync (MS_ASYNC) is not supported so all syncs are
-    // treated as though they were non-asynchronous (MS_SYNC). The caller will
-    // be unable to detect any difference in behavior.
-
-    // Note: experimentation reveals that Linux invalidates other mappings
-    // of the same file, whether  MS_INVALIDATE is present or not (meaning
-    // that they too are updated to reflect the contents of the file with
-    // or without the MS_INVALIDATE flag).
+    ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+    ECHECK((index = _get_page_index(addr, length)));
 
-    myst_spin_lock(&_msync_mappings_lock);
+    myst_rspin_lock(&_mman.lock);
     {
-        /* flush any msync mappings contained by this address range */
-        for (msync_mapping_t* p = _msync_mappings; p; p = p->next)
+        const size_t n = length / PAGE_SIZE;
+        const uint8_t* page = addr;
+
+        for (size_t i = index; i < index + n; i++)
         {
-            uint8_t* lo = addr;
-            uint8_t* hi = (uint8_t*)addr + length;
-            uint8_t* plo = p->addr;
-            uint8_t* phi = (uint8_t*)p->addr + p->length;
-            uint8_t* maxlo = _max_ptr(lo, plo);
-            uint8_t* minhi = _min_ptr(hi, phi);
-
-            // [AAAAAAAAAAAAAAAAA]
-            //       [MMMMMMMMMMMMMMMMMMMMM]
-            //       ^           ^
-            //      maxlo      minhi
-            //
-            //       [AAAAAAAAAAAAAAAAA]
-            // [MMMMMMMMMMMMMMM]
-            //       ^         ^
-            //      maxlo     minhi
-
-            if (maxlo < minhi)
-            {
-                ECHECK(_sync_file(
-                    p->fd,                     /* fd */
-                    p->offset + (maxlo - plo), /* offset */
-                    maxlo,                     /* addr */
-                    minhi - maxlo));           /* length */
-            }
-        }
+            myst_fdmapping_t* p = &v.fdmappings[i];
+
+            if (p->used == MYST_FDMAPPING_USED)
+                ECHECK(_sync_file(p->fd, p->offset, page, PAGE_SIZE));
 
-        // ATTN: currently msync() does not update other mappings of the same
-        // file. This would involve refreshing those mappings from the file
-        // blocks. This case may be rare since it would require mapping the
-        // same file to different memory regions.
+            page += PAGE_SIZE;
+        }
     }
-    myst_spin_unlock(&_msync_mappings_lock);
+    myst_rspin_unlock(&_mman.lock);
 
 done:
     return ret;
 }
 
-/* notified on close to remove msync mappings involving fd */
+/* notified on close to clear msync mappings involving fd */
```

Comment:
```suggestion
/* notified on close to clear fdmappings involving fd */
```

Reply:
This diff had a conflicts but I manually applied it. Thanks!